llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp

```diff
@@ -3170,13 +3169,17 @@ static bool isUsedByLoadStoreAddress(const VPUser *V) {
       }
     }
 
-    // The legacy cost model only supports scalarization loads/stores with phi
-    // addresses, if the phi is directly used as load/store address. Don't
-    // traverse further for Blends.
-    if (Blend)
-      continue;
-
-    append_range(WorkList, Cur->users());
+    for (VPUser *U : Cur->users()) {
+      // The legacy cost model only supports scalarization loads/stores with phi
+      // addresses, if the phi is directly used as load/store address. Don't
+      // traverse further for PHI selects.
+      if (isa_and_nonnull<PHINode>(Cur->getUnderlyingValue()) &&
+          (!isa<VPSingleDefRecipe>(U) ||
+           cast<VPSingleDefRecipe>(U)->getUnderlyingValue() !=
+               Cur->getUnderlyingValue()))
+        continue;
```

Comment:
Not sure why there is a reference to loads/stores in the comment, but the code seems to use VPSingleDefRecipe? Not sure I understand this.